packages/teleport/src/components/TdpClientCanvas/TdpClientCanvas.tsx

```diff
@@ -32,135 +37,234 @@ export default function TdpClientCanvas(props: Props) {
-        if (onMouseWheelScroll) canvas.removeEventListener('wheel', _onwheel);
+        tdpCli.removeListener(TdpClientEvent.WS_OPEN, tdpCliOnWsOpen);
       };
     }
-  }, [tdpCli]);
+  }, [tdpCli, tdpCliOnWsOpen]);
+
+  useEffect(() => {
+    const canvas = canvasRef.current;
+    const _oncontextmenu = onContextMenu;
+    if (onContextMenu) {
+      canvas.oncontextmenu = _oncontextmenu;
+    }
+
+    return () => {
+      if (onContextMenu)
+        canvas.removeEventListener('contextmenu', _oncontextmenu);
+    };
+  }, [onContextMenu]);
+
+  useEffect(() => {
+    const canvas = canvasRef.current;
+    const _onmousemove = (e: MouseEvent) => {
+      onMouseMove(tdpCli, canvas, e);
+    };
+    if (onMouseMove) {
+      canvas.onmousemove = _onmousemove;
+    }
+
+    return () => {
+      if (onMouseMove) canvas.removeEventListener('mousemove', _onmousemove);
+    };
+  }, [onMouseMove]);
+
+  useEffect(() => {
+    const canvas = canvasRef.current;
+    const _onmousedown = (e: MouseEvent) => {
+      onMouseDown(tdpCli, e);
+    };
+    if (onMouseDown) {
+      canvas.onmousedown = _onmousedown;
+    }
+
+    return () => {
+      if (onMouseDown) canvas.removeEventListener('mousedown', _onmousedown);
+    };
+  }, [onMouseDown]);
+
+  useEffect(() => {
+    const canvas = canvasRef.current;
+    const _onmouseup = (e: MouseEvent) => {
+      onMouseUp(tdpCli, e);
+    };
+    if (onMouseUp) {
+      canvas.onmouseup = _onmouseup;
+    }
+
+    return () => {
+      if (onMouseUp) canvas.removeEventListener('mouseup', _onmouseup);
+    };
+  }, [onMouseUp]);
+
+  useEffect(() => {
+    const canvas = canvasRef.current;
+    const _onwheel = (e: WheelEvent) => {
+      onMouseWheelScroll(tdpCli, e);
+    };
+    if (onMouseWheelScroll) {
+      canvas.onwheel = _onwheel;
+    }
+
+    return () => {
+      if (onMouseWheelScroll) canvas.removeEventListener('wheel', _onwheel);
+    };
+  }, [onMouseWheelScroll]);
+
+  useEffect(() => {
+    const canvas = canvasRef.current;
+    const _onkeydown = (e: KeyboardEvent) => {
+      onKeyDown(tdpCli, e);
+    };
+    if (onKeyDown) {
+      canvas.onkeydown = _onkeydown;
+    }
+
+    return () => {
+      if (onKeyDown) canvas.removeEventListener('keydown', _onkeydown);
+    };
+  }, [onKeyDown]);
+
+  useEffect(() => {
+    const canvas = canvasRef.current;
+    const _onkeyup = (e: KeyboardEvent) => {
+      onKeyUp(tdpCli, e);
+    };
+    if (onKeyUp) {
+      canvas.onkeyup = _onkeyup;
+    }
+
+    return () => {
+      if (onKeyUp) canvas.removeEventListener('keyup', _onkeyup);
+    };
+  }, [onKeyUp]);
+
+  useEffect(() => {
+    const canvas = canvasRef.current;
+    const _onmouseenter = (e: MouseEvent) => {
+      onMouseEnter(tdpCli, e);
+    };
+    if (onMouseEnter) {
+      canvas.onmouseenter = _onmouseenter;
+    }
+
+    return () => {
+      if (onMouseEnter)
+        canvas.removeEventListener('onmouseenter', _onmouseenter);
+    };
+  }, [onMouseEnter]);
```

Comment:
hm... i think all these useeffect's will run every re-render regardless if it really changed or not, b/c functions get redefined per re-render (they get a new reference) so to react they look different. You can test it out with console log and see if these really run every time this component re-renders.